### ISSUE TITLE
Iroha core's `instructions_not_allowed` and `queries_not_allowed` tests not functioning as intended

### ISSUE THREAD

**Issue by DCNick3** [hyperledger/iroha#4821]
# TL;DR

Even though these tests pass, I believe this was not intended.

https://github.com/hyperledger/iroha/blob/b415c73ca5ce8d3258ca0b0bf188e9d54a9010e6/core/src/smartcontracts/wasm.rs#L1957

Here^ it is checked that if the error is a certain type, and if so - some additional asserts are added about the contents. But if the error is of different type, the tests silently passes, without reporting the error.

Coincidentally, both of those tests are in actuality failing due to a `Instantiation(InstantiationError::Linker(...))` error, probably because of some changes to export names/signatures.

My attempts to fix the linking errors led the query execution being successful, even though the tests wants it to fail. I am not sure how to properly fix these, hence this issue.
